src/backends/netbsd.rs

```diff
@@ -42,36 +38,24 @@ unsafe extern "C" fn polyfill_using_kern_arand(
 
 type GetRandomFn = unsafe extern "C" fn(*mut c_void, libc::size_t, libc::c_uint) -> libc::ssize_t;
 
-static GETRANDOM: AtomicPtr<c_void> = AtomicPtr::new(ptr::null_mut());
-
 #[cold]
 #[inline(never)]
-fn init() -> *mut c_void {
-    static NAME: &[u8] = b"getrandom\0";
-    let name_ptr = NAME.as_ptr().cast::<libc::c_char>();
-    let mut ptr = unsafe { libc::dlsym(libc::RTLD_DEFAULT, name_ptr) };
+fn init() -> usize {
+    let mut ptr = unsafe { libc::dlsym(libc::RTLD_DEFAULT, c"getrandom".as_ptr()) };
     if ptr.is_null() || cfg!(getrandom_test_netbsd_fallback) {
         // Verify `polyfill_using_kern_arand` has the right signature.
         const POLYFILL: GetRandomFn = polyfill_using_kern_arand;
         ptr = POLYFILL as *mut c_void;
     }
-    GETRANDOM.store(ptr, Ordering::Release);
-    ptr
+    ptr as usize
 }
 
 #[inline]
 pub fn fill_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
-    // Despite being only a single atomic variable, we still cannot always use
-    // Ordering::Relaxed, as we need to make sure a successful call to `init`
-    // is "ordered before" any data read through the returned pointer (which
-    // occurs when the function is called). Our implementation mirrors that of
-    // the one in libstd, meaning that the use of non-Relaxed operations is
-    // probably unnecessary.
-    let mut fptr = GETRANDOM.load(Ordering::Acquire);
-    if fptr.is_null() {
-        fptr = init();
-    }
-    let fptr = unsafe { mem::transmute::<*mut c_void, GetRandomFn>(fptr) };
+    static GETRANDOM_FN: lazy::LazyUsize = lazy::LazyUsize::new();
+
+    let fptr = GETRANDOM_FN.unsync_init(init);
+    let fptr = unsafe { mem::transmute::<usize, GetRandomFn>(fptr) };
```

Comment:
As @newpavlov mentioned in https://github.com/rust-random/getrandom/pull/758#discussion_r2557703093 this might be wrong w.r.t. pointer-provenance. See https://doc.rust-lang.org/std/ptr/index.html#provenance

That document explicitly recommends using `AtomicPtr` over `AtomicUsize` for the sorts of manipulations here. We used to have code like what you are suggesting here, but we changed it to its current version to:
  - Align with what's in the standard library (see [`DlsymWeak` in `std`](https://github.com/rust-lang/rust/blob/b64df9d1012f2482b54a4d959548cf8fc67e820c/library/std/src/sys/pal/unix/weak/dlsym.rs#L28))
  - Ensure correct pointer provenance

Reply:
Good chance I'm being dense, but what is the provenance problem here? The point of provenance is to say that a point is not just a number, it also carries temporal and mutability "metadata" (conceptually, not physically). The mere act of casting to an integer does not violate provenance.

The vast majority of the discussion in that document (including the suggestion to use AtomicPtr) is focused on mutability, which doesn't apply here because we do not write through this pointer. We also do not offset it or otherwise derive a new address from it on which we need to retain the same provenance.

Reply:
You lose pointer provenance in naive pointer-integer-pointer casts. While it works fine in practice, it's still recommended to avoid it whenever possible.